### ISSUE TITLE
Use a unique payload for every run

### ISSUE THREAD

**Issue by cp-k1nakayama** [alexcasalboni/aws-lambda-power-tuning#87]
Hi @alexcasalboni 

#85  issue didn't seem to convey my problem properly.

Currently, when 6 types of powerValues ​​are given and 5 is given to num, the Lambda Function is invoked 30 times in total.
If this Lambda Function behaves like deleting one record of the ID passed by Payload for each Invoke, the type of Payload needs to prepare more than the total number of calls.

However, according to the current specifications, it is not possible to give more payloads than the value of num.

Therefore, when the execution of the first PowerValues ​​ends, the record with the ID specified in Payload has been deleted from the DB, and the specified ID does not exist when the second and subsequent PowerValues ​​are executed. The correct processing is not performed and the required PowerValue cannot be checked.

In order to solve this problem, it is necessary to modify Payload so that it can be given more than the total number of executions.

Alternatively, by implementing a function that allows you to specify a Lambda Function that performs processing before and after each execution, you can execute it by giving only one type of payload. I think that such a function is very useful.

**Comment by alexcasalboni** [comment]
Hi @cp-k1nakayama, I see what you mean now.

Actually the different power values are running in parallel (always, even with `parallelInvocation: false`).

I like the idea of a pre/post-processing function, if we manage to not complicate the whole state macine and make it more expensive to run too even for simple cases.

Let me think of a couple of potential solutions that allows us to cover this case, as well as any other function with side effects.

## Solution 1 (no pre/post-processing)

I was thinking of a "simple" payload variable concept, that'd allow you to inject dynamic values during the execution. The input could look like this: 

```json
{
    "lambdaARN": "your-lambda-function-arn",
    "powerValues": [128, 256, 512, 1024, 2048, 3008],
    "num": 5,
    "payload": {
        "endpoint": "/abc/def/${ID}"
    },
    "payloadVariables": {
        "ID": ["a", "b", "c", "d", "e"]
    }
}
```

This way you'd be able to specify a variable in your payload (`${ID}`) and also provide a set of 5 values (because `num: 5`) in the new `payloadVariables` input attribute.

The state machine will invoke your function 5 times, each time with a different payload.

We would still have a problem with the parallel execution of the 6 power values. So we could say that in `payloadVariables` you can either set 5 values (used for all power values) or 30 values (6 times 5, so that we have enough unique values overall).

You would still need to prepare these 30 unique records in your database prior to the state machine execution.


## Solution 2 (with pre/post-processing)

```json
{
    "lambdaARN": "your-lambda-function-arn",
    "powerValues": [128, 256, 512, 1024, 2048, 3008],
    "num": 5,
    "payload": {...},
    "preProcessingFunction": "your-lambda-function-arn",
    "postProcessingFunction": "your-lambda-function-arn"
}
```

This way, we don't have to change the state machine structure and you'd be responsible for implementing the pre/post-processing functions (only one, or both).

In your case, you'd need to plug a pre-processing function that will create a new unique record to be deleted. Keep in mind that the different power values will still be tested in parallel, so your pre-processing function might be invoked concurrently (up to 6 times if `parallelInvocation: false` and up to 30 times if `parallelInvocation: true`).

We could also agree that the pre-processing function's output (if provided) will be used as tuning payload, while the post-processing function will receive in input your function's output (`pre-processing -> function -> post-processing`). Makes sense?

---

Which of these two alternatives do you prefer?

Personally, I think the 2nd option is much more flexible and easy to understand, even if it forces you to provision additional Lambda functions before you can run the state machine.

If you're happy with option 2, I'll happy to proceed with a first draft of the implementation soon.

**Comment by cp-k1nakayama** [comment]
Thank you very much.

Certainly, I think either of these methods can solve the problem.
However, assuming that CI / CD continuously executes these, in the case of the Solution 1 method, it is necessary to separately implement preprocessing and unnecessary data may remain.

For those reasons, I like Solution 2.

Also, the idea that the contents output in the pre-processing for the payload can be used is very flexible and good.
In this case, I think it is good to be able to use the value given to the Payload parameter as the Payload of the preprocessing.
(If i can use this method, i can reuse the pre-processing and post-processing Lambda Function when executing various functions.)

**Comment by alexcasalboni** [comment]
Great, thanks for sharing your feedback :)

I'll start working on this asap.

**Comment by alexcasalboni** [comment]
I've drafted a first version of the implementation with same basic tests in #88 :)

I won't be able to finalize it until another PR is merged as well, but feel free to provide feedback or even deploy and test that branch if you're feeling brave 💪

**Comment by alexcasalboni** [comment]
@cp-k1nakayama FYI the PR branch is now stable and almost ready to be merged (only a documentation update is missing). Feel free to give it a try and share feedback.

I'll try to merge and publish a new SAR app version later this week 🎉

**Comment by alexcasalboni** [comment]
Merged 🚀
